proposals/features-library.md

```diff
@@ -0,0 +1,76 @@
+# Sharing code between Dev Container Features
+
+ref: https://github.com/devcontainers/spec/issues/129
+
+# Motivation
+
+The current model for authoring Features encourages significant copy and paste between individual Features.  Most of the Features published under the `devcontainers/features` namespace contain approximately 100 lines of [identifical helper functions](https://github.com/devcontainers/features/blob/main/src/azure-cli/install.sh#L29-L164) before the first line of the script is executed.  The process of maintaining these helper functions is tedious, error-prone, and leads to less readable Features.
+
+Additionally, [members of the Feature authoring community have expressed interest in sharing code between Features](https://github.com/orgs/devcontainers/discussions/10).  Exploring the source code of self-published Features, it is evident that many authors have copy/pasted code from other Features.  This is a sign that there is a need for a more efficient way to share code between Features.
+
+# Goal
+
+Provide a generic and efficient way to share code between Features.
+
+# Proposal
+
+Add an `include` property to the `devcontainer-feature.json`.  This will be honored during the Feature [packaging stage](https://containers.dev/implementors/features-distribution/#packaging). The contents of the provided relative path is copied into the packaged Feature archive.
+
+Property | Type | Description
+--- | --- | ---
+`include` | `string[]` | An array of relative paths to package with the Feature. Paths are relative to root directory when packaging (often the git root directory).  If the element is a folder, it is copied recursively.
+
+## Example
+
+```json
+{
+    "name": "my-feature",
+    "version": "0.0.1",
+    "include": [
+        "utils/",
+    ]
+}
+```
+
+The preceeding example will recursively copy the contents of the the `utils` folder into the temporary folder used to package the Feature.  The `utils` folder will be included with the published Features.
+
+This functionality will enable sharing a common set of helper functions between Features.  The following example shows how a Feature could include a `utils` folder that contains a `common.sh` file.
+
+```bash
+# utils/common.sh
+function common_function() {
+   echo "Hello, '$1'"
+}
+```
+
+```bash
+# my-feature/install.sh
+
+# Include common functions
+source "./utils/common.sh"
```

Comment:
I guess this is `./utils` and not `../utils` because the packaged folder structure is different from the folder structure in the repository? So this can't be run from within the repository because the paths do not match?

Wouldn't it make development of features easier if the folder structures were the same? E.g., can we make it such that features can be run from the repository as "local" features during feature development?

@samruddhikhandale Any insights from your work on features?

Reply:
Yes, that's how I was originally imagining it.  Anything in the `includes` would be copied recursively to the root of the Feature's packaged directory.

Reply:
I would suggest to keep the folder structure the same as in the repository. The main difficulty from what I see is the lookup of the `devcontainer-feature.json` which is currently expected at the root of the archive. If we know the feature id, we could additionally check in the folder with the same name as the unqualified id.

Reply:
For the Feature development case I don't think that makes too much of a difference - the `features test` command currently does a full copy into a temp folder inside of a temporary project's `.devcontainer` (to make it a 'local Features' from the perspective of the CLI).  This could likely be optimized, but we have strict rules over where local Features can live relative to the devcontainer.json, so it's a bit of hassle to _not_ do the copying.  I see the appeal of keeping the directory structure, and depending on if a 'local' Feature or a 'packaged' Feature, change where we look for the `include` files.  That said, we haven't really gotten feedback that performance on the testing command is an issue, and the vast majority of the time is not spent setting up the tests.

I don't know how useful that behavior is in practice, and makes figuring out the correct path slightly more confusing when trying to use an included script in a packaged Feature.

Also, how would it work sourcing from `../utils`.  Would we need to re-adjust how the entire archive is structured (ie: create a dummy `src` folder that has one folder `feature-name`, that contains `install.sh` and `devcontainer-feature.json`?